### DISCUSSION TITLE
Looking for an example to bind std::deque, Thanks!

### DISCUSSION THREAD

**Discussion by ffyu-lj** [pybind/pybind11#4292] [Q&A] [answered]
Example:
```
struct SampleType
{
    int id;
    string name;
}
std::deque<SampleType> Deque_ST;
```
Bind:
```
PYBIND11_MAKE_OPAQUE(std::deque<SampleType>)

py::bind_deque(Deque_ST)(m, "Deque_ST")
```
But I got error: "error: 'bind_deque' is not a member of 'py'."

I thought binding deque should be similar to binding vector or map (py::bind_vector<std::vector<int>).
but it doesn't seem so, or I did something wrong?
Please kindly share your thoughts.
Thanks.

**Comment by Skylion007** [comment, marked as answer]
Assuming you have a caster made for sampel `SampleType`, you should include `pybind/stl.h` and that will give you a std::deque caster automatically. Alternatively, `std::deque<std::pair<int,std::string>>` should work out of the box as a non-opaque caster.

We don't have any casters for std::deque written in stl_bind.h. STL bind_vector and bind_map are the only two implement in stl_bind.h. Otherwise, you would need to bind all the methods yourself manually.

**Comment by ffyu-lj** [comment]
@Skylion007 , thank you for your attention.

You mean std::deque<SampleType> should work out of box, correct?

I've tried that (no binding code for it), but when I run my test script ( the std::deque<SampleType> is a return type for another function), I got error saying " TypeError: Unable to convert runction return value to a Python type! The signature was\n\t() -> List[namespace::SampleType]\n\n At:\n  /tmp/test.py(472): func_test1\n\n"

**Comment by Skylion007** [comment]
@ffyu-lj Not unless you define a caster for SampleType. If it was an std::tuple or std::pair, it woudl work. Or you need to go and create a custom caster for SampleType or bind SampleType as a Python class

**Reply by ffyu-lj** [reply]
I did bind SampleType as a python class, but it was in another module. 
My guess is that SampleType and std::deque<SampleType> should be in the same module.

**Reply by ffyu-lj** [reply]
They are in different modules because it is a big project (modules are decided where the variables are defined).
